apps/marketing/src/components/(marketing)/single-player-mode/create-single-player-document.action.ts

```diff
@@ -0,0 +1,220 @@
+
+    // Fetch service user who will be the owner of the document.
+    const serviceUser = await tx.user.findFirstOrThrow({
+      where: {
+        email: SERVICE_USER_EMAIL,
+      },
+    });
+
+    // Create document and recipient.
+    const document = await tx.document.create({
+      data: {
+        title: documentName,
+        status: DocumentStatus.COMPLETED,
+        userId: serviceUser.id,
+        document: Buffer.from(pdfBytes).toString('base64'),
+        created: createdAt,
+        Recipient: {
+          create: {
+            name: signer.name,
+            email: signer.email,
+            token: documentToken,
+            signedAt: createdAt,
+            readStatus: ReadStatus.OPENED,
+            signingStatus: SigningStatus.SIGNED,
+            sendStatus: SendStatus.SENT,
+          },
+        },
+      },
+    });
+
+    // Create fields and signatures.
+    await Promise.all(
+      fields.map((field) =>
+        tx.field.create({
+          data: {
+            documentId: document.id,
+            ...mapFields(field, signer),
+            Signature:
+              field.type === FieldType.SIGNATURE
+                ? {
+                    create: {
+                      Recipient: {
+                        connect: {
+                          documentId_email: {
+                            email: signer.email,
+                            documentId: document.id,
+                          },
+                        },
+                      },
+                      signatureImageAsBase64: signatureImageAsBase64,
+                      typedSignature,
+                    },
+                  }
+                : undefined,
+          },
+        }),
+      ),
+    );
+
+    return documentToken;
+  });
+
+  // Todo: Handle `downloadLink` and `reviewLink`.
+  const template = createElement(DocumentSelfSignedEmailTemplate, {
+    downloadLink: 'https://documenso.com',
+    reviewLink: 'https://documenso.com',
+    documentName: documentName,
+    assetBaseUrl: process.env.NEXT_PUBLIC_SITE_URL || 'http://localhost:3000',
+  });
+
+  // Send email to signer.
+  await mailer.sendMail({
+    to: {
+      address: signer.email,
+      name: signer.name,
+    },
+    from: {
+      name: FROM_NAME,
+      address: FROM_ADDRESS,
+    },
+    subject: 'Document signed',
+    html: render(template),
+    text: render(template, { plainText: true }),
+  });
+
+  return documentToken;
+};
+
+/**
+ * Map the fields provided by the user to fields compatible with Prisma.
+ *
+ * Signature fields are handled separately.
+ *
+ * @param field The field passed in by the user.
+ * @param signer The details of the person who is signing this document.
+ * @returns A field compatible with Prisma.
+ */
+const mapFields = (
+  field: TCreateSinglePlayerDocumentSchema['fields'][number],
+  signer: TCreateSinglePlayerDocumentSchema['signer'],
+) => {
+  let customText = '';
+
+  switch (field.type) {
```

Comment:
Why not use `ts-pattern` for this? 🤔

Reply:
I was thinking that ts-pattern would mainly be used for templates or more complicated logic like that big query you did.

Would it be a little overkill for this?

Reply:
It might be but switch statements tilt me, it also lets us remove a `let` in favour of 
```
const customText = match(X).with(...).with(...).otherwise(...)
```

Reply:
Alright i'll toss it in, i'm always down to try new stuff 😄